march_gazebo_plugins/src/stairs_controller.cpp

```diff
@@ -0,0 +1,92 @@
+// Copyright 2019 Project March.
+
+#include <stairs_controller.h>
+#include <gazebo/physics/physics.hh>
+
+namespace gazebo
+{
+StairsController::StairsController(physics::ModelPtr model) : ObstacleController(model)
+{
+  this->swing_step_size_ = 0.7;  // This estimate will be adjusted every step
+
+  this->p_pitch_ = 60;
+  this->d_pitch_ = 10;
+  this->p_roll_ = 150;
+  this->d_roll_ = 0;
+  this->p_yaw_ = 500;
+  this->d_yaw_ = 25;
+
+  this->error_x_last_timestep_ = 0;
+  this->error_y_last_timestep_ = 0;
+  this->error_yaw_last_timestep_ = 0;
+}
+
+void StairsController::update(ignition::math::v4::Vector3<double>& torque_all,
+                              ignition::math::v4::Vector3<double>& torque_stable)
+{
+  // Note: the exo moves in the negative x direction, and the right leg is in
+  // the positive y direction
+
+  double time_since_start = this->model_->GetWorld()->SimTime().Double() - this->subgait_start_time_;
+  if (time_since_start > this->subgait_duration_)
+  {
+    this->subgait_name_ = "home_stand";
+    this->subgait_changed_ = true;
+  }
+
+  auto model_com = this->GetCom();
+
+  // Left foot is stable unless subgait name starts with left
+  auto stable_foot_pose = this->foot_left_->WorldCoGPose().Pos();
+  std::string stable_side = "left";
+  if (this->subgait_name_.substr(0, 4) == "left")
+  {
+    stable_foot_pose = this->foot_right_->WorldCoGPose().Pos();
+    stable_side = "right";
```

Comment:
May be you could check whether the gait starts with right and otherwise throw a warning.

Reply:
Okay